File: src/config/unitConversion.js
/**
 * UNIT CONVERSION — §10.2 of the master spec.
 *
 * Millimetres are NEVER used as raw Three.js units. Scene units are METRES.
 * Every mesh in `components/digitalTwin` sizes itself through this module, so
 * there is exactly one place where industrial units become scene units.
 */
import { GRAVITY } from './engineeringConfig';
import { millConfig } from './millConfig';
/** 1 scene unit = 1 metre. */
export const SCENE_UNITS_PER_METRE = 1;
/** Industrial millimetres -> scene units. */
export function mmToScene(mm) {
    return (mm / 1000) * SCENE_UNITS_PER_METRE;
}
/** Scene units -> industrial millimetres. */
export function sceneToMm(scene) {
    return (scene / SCENE_UNITS_PER_METRE) * 1000;
}
/** Industrial metres -> scene units. */
export function mToScene(m) {
    return m * SCENE_UNITS_PER_METRE;
}
/**
 * Strip thickness -> scene units, with the disclosed visual exaggeration (§10.3).
 * The displayed NUMBER is always the true thickness; only the geometry is scaled.
 */
export function stripThicknessToScene(thicknessMm) {
    return mmToScene(thicknessMm * millConfig.visual.stripThicknessExaggeration);
}
/**
 * Work-roll surface separation for a given roll gap (§10.3):
 *   workRollDistance = stripThickness(exaggerated) + rollGapVisualOffset
 */
export function rollGapToScene(rollGapMm) {
    return (mmToScene(rollGapMm * millConfig.visual.rollGapExaggeration) +
        millConfig.visual.rollGapVisualOffset);
}
// ---------------------------------------------------------------------------
// Engineering unit conversions
// ---------------------------------------------------------------------------
/** Metric tonnes-force -> kilonewtons. */
export function tonnesToKN(t) {
    return t * GRAVITY;
}
/** Kilonewtons -> metric tonnes-force. */
export function kNToTonnes(kN) {
    return kN / GRAVITY;
}
/** Newtons -> metric tonnes-force. */
export function newtonsToTonnes(n) {
    return n / (GRAVITY * 1000);
}
/** m/min -> m/s. */
export function mpmToMps(mpm) {
    return mpm / 60;
}
/** m/s -> m/min. */
export function mpsToMpm(mps) {
    return mps * 60;
}
/** Millimetres -> micrometres. */
export function mmToUm(mm) {
    return mm * 1000;
}
/** Micrometres -> millimetres. */
export function umToMm(um) {
    return um / 1000;
}
/** rpm -> radians per second. */
export function rpmToRadPerSec(rpm) {
    return (rpm * 2 * Math.PI) / 60;
}
/** radians per second -> rpm. */
export function radPerSecToRpm(rad) {
    return (rad * 60) / (2 * Math.PI);
}
// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------
export function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
}
/** Linear interpolation. */
export function lerp(a, b, t) {
    return a + (b - a) * t;
}
/**
 * Frame-rate independent exponential damping (§10.4).
 * `halfLife` is the time in seconds for the remaining error to halve, so the
 * visual response is identical at 30 fps and 144 fps.
 */
export function damp(current, target, halfLife, dt) {
    if (halfLife <= 0)
        return target;
    const factor = 1 - Math.pow(2, -dt / halfLife);
    return current + (target - current) * factor;
}
/** Format a number for a control-room readout: fixed decimals, no locale commas. */
export function fmt(value, decimals = 1) {
    if (value === null || value === undefined || !Number.isFinite(value))
        return '—';
    return value.toFixed(decimals);
}

File: dotnet/src/Crm04.Domain/Simulation/UnitConversion.cs
using Crm04.Domain.Configuration;

namespace Crm04.Domain.Simulation;

/// <summary>
/// UNIT CONVERSION - §10.2 of the master spec. Port of <c>src/config/unitConversion.ts</c>.
///
/// SCOPE NOTE. The TypeScript module also carries the scene-unit helpers (<c>mmToScene</c>,
/// <c>rollGapToScene</c>, <c>stripThicknessToScene</c>) because the React twin imported its
/// geometry conversions from the same file. Those are NOT ported here: in the .NET topology the
/// scene lives in <c>wwwroot/js/twin/</c> and owns its own conversions, and a Domain project that
/// knew about Three.js units would be the wrong shape. What is ported is the engineering half -
/// the conversions the physics models actually call.
/// </summary>
public static class UnitConversion
{
    /// <summary>Metric tonnes-force -&gt; kilonewtons.</summary>
    public static double TonnesToKn(double t) => t * EngineeringConfig.Gravity;

    /// <summary>Kilonewtons -&gt; metric tonnes-force.</summary>
    public static double KnToTonnes(double kN) => kN / EngineeringConfig.Gravity;

    /// <summary>Newtons -&gt; metric tonnes-force.</summary>
    public static double NewtonsToTonnes(double n) => n / (EngineeringConfig.Gravity * 1000);

    /// <summary>m/min -&gt; m/s.</summary>
    public static double MpmToMps(double mpm) => mpm / 60;

    /// <summary>m/s -&gt; m/min.</summary>
    public static double MpsToMpm(double mps) => mps * 60;

    /// <summary>Millimetres -&gt; micrometres.</summary>
    public static double MmToUm(double mm) => mm * 1000;

    /// <summary>Micrometres -&gt; millimetres.</summary>
    public static double UmToMm(double um) => um / 1000;

    /// <summary>rpm -&gt; radians per second.</summary>
    public static double RpmToRadPerSec(double rpm) => rpm * 2 * Math.PI / 60;

    /// <summary>radians per second -&gt; rpm.</summary>
    public static double RadPerSecToRpm(double rad) => rad * 60 / (2 * Math.PI);

    /// <summary>
    /// Clamp. Written as the TypeScript writes it - <c>min(max(v, lo), hi)</c> - rather than as
    /// <see cref="Math.Clamp(double,double,double)"/>, which throws when <c>lo &gt; hi</c> where
    /// JavaScript quietly returns <c>hi</c>. The models never pass an inverted range, but the
    /// difference is a throw versus a value and that is not a difference to leave to chance.
    /// </summary>
    public static double Clamp(double value, double min, double max) =>
        Math.Min(Math.Max(value, min), max);

    /// <summary>Linear interpolation.</summary>
    public static double Lerp(double a, double b, double t) => a + (b - a) * t;

    /// <summary>
    /// Frame-rate independent exponential damping (§10.4). <paramref name="halfLife"/> is the time
    /// in seconds for the remaining error to halve, so the response is identical at any tick rate.
    /// </summary>
    public static double Damp(double current, double target, double halfLife, double dt)
    {
        if (halfLife <= 0) return target;
        var factor = 1 - Math.Pow(2, -dt / halfLife);
        return current + (target - current) * factor;
    }
}
